features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/ecp/ecp_internal_alt.c

```diff
@@ -0,0 +1,828 @@
+ *
+ * \return          0 if successful.
+ */
+int mbedtls_internal_ecp_normalize_jac_many(const mbedtls_ecp_group *grp,
+                                            mbedtls_ecp_point *T[], size_t t_len)
+{
+    if (T == NULL || t_len == 0) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    int ret;
+    
+    mbedtls_ecp_point **ecp_point = T;
+    mbedtls_ecp_point **ecp_point_end = T + t_len;
+    
+    for (; ecp_point != ecp_point_end; ecp_point ++) {
+        MBEDTLS_MPI_CHK(mbedtls_internal_ecp_normalize_jac(grp, *ecp_point));
+    }
+
+cleanup:
+
+    return ret;
+}
+#endif
+
+int mbedtls_internal_run_eccop(const mbedtls_ecp_group *grp,
+                            mbedtls_ecp_point *R,
+                            const mbedtls_mpi *m,
+                            const mbedtls_ecp_point *P,
+                            const mbedtls_mpi *n,
+                            const mbedtls_ecp_point *Q,
+                            uint32_t eccop)
+{
+    /* Check necessary arguments for all ECC operations */
+    if (grp == NULL || R == NULL) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    /* Check grp->P is positive */
+    if (mbedtls_mpi_cmp_int(&grp->P, 0) <= 0) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    /* Check supported maximum key bits */
+    if (grp->pbits > NU_ECC_MAXKEYBITS) {
+        return MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+    }
+
+    int ret;
+    mbedtls_mpi N_;
+    const mbedtls_mpi *Np;
+    
+    mbedtls_mpi_init(&N_);
+    
+    /* Use MBEDTLS_INTERNAL_MPI_NORM(Np, N1, N_, P) to get normalized MPI
+     *
+     * N_: Holds normalized MPI if the passed-in MPI N1 is not
+     * Np: Pointer to normalized MPI, which could be N1 or N_
+     */
+    
+    /* Check necessary arguments and handle special cases for specified ECC operation 
+     *
+     * ECCOP_POINT_MUL      R = m*P
+     * ECCOP_POINT_ADD      R = P + Q
+     * ECCOP_POINT_DOUBLE   R = 2*P
+     *
+     * ECC accelerator doesn't support R = 0, and we need to detect it.
+     */
+    if (eccop == ECCOP_POINT_MUL) {
+        /* R = m*P */
+        if (m == NULL || P == NULL) {
+            ret = MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+        
+         /* R = 0*P = 0 or R = P = 0 */
+        if (mbedtls_mpi_cmp_int(m, 0) == 0 || mbedtls_mpi_cmp_int(&P->Z, 0) == 0) {
+            ret = mbedtls_ecp_set_zero(R);
+            goto cleanup;
+        }
+        
+        /* R = 1*P */
+        if (mbedtls_mpi_cmp_int(m, 1) == 0) {
+            MBEDTLS_MPI_CHK(mbedtls_ecp_copy(R, P));
+            MBEDTLS_MPI_CHK(mbedtls_internal_ecp_normalize_jac(grp, R));
+            goto cleanup;
+        }
+        
+        /* R = m*P = (multiple of order)*G = 0 */
+        /* NOTE: If grp->N (order) is a prime, we could detect R = 0 for all m*P cases
+         *       by just checking if m is a multiple of grp->N. Otherwise, sigh. */
```

Comment:
The order `grp->N` is not the order of the entire elliptic curve, but the order of the subgroup `<G>` spanned by `ecp->G`, which is assumed to be prime. Therefore the check is fine and exhausts `R=0` completely, _provided_ one makes the assumption that `P` lies in `<G>` -- whereas without this assumption, the check doesn't make sense in the first place. I think `P` belonging to `<G>` is a reasonable yet undocumented assumption the default S/W ECC code also makes at times, but I don't remember details - ping @mpg? Continuing on this assumption, the case of order `2`, i.e. `P = (-P)` but `P != 0`, cannot occur.

What happens to the H/W accelerator if `R=0` is not covered beforehand? It seems that it's not capable of returning non-affine points, or is it?

Reply:
For the short Weierstrass case, for all curves we currently support, `<G>` is the whole curve, and `mbedtls_ecp_check_pubkey()` (called at the beginning of `mbedtls_ecp_mul()` and hopefully all other public functions that receive both secret input and some untrusted input) ensures that the point lies on the curve, hence that it belongs to `<G>`. If we ever added support for short weierstrass curves with non-trivial cofactor, we would need to adapt `mbedtls_ecp_check_pubkey()` so that it keeps checking that the point is a multiple of `G`.

Also, `mbedtls_ecp_check_privkey()` ensures that `m` is in the range [1, N) so together with the previous check on `N`, this ensures that `m * P != 0`. Note however that this protection only applies to the scalar multiplication part: when doing the final addition in `R = m * P + n * Q` you still need to be prepared for the `R == 0` case.

Reply:
> What happens to the H/W accelerator if R=0 is not covered beforehand? It seems that it's not capable of returning non-affine points, or is it?

@hanno-arm If `R = 0`, H/W accelerator will hang and its finish/error flag won't be set. That's why I must filter out all `R = 0` cases.